File: smoke_tests/lua/log_injection.lua
-- Log Injection vulnerabilities in Lua

-- Test 1: Basic log injection with print
function log_user_action(username, action)
    -- VULNERABLE: User input in log
    print("[INFO] User " .. username .. " performed action: " .. action)
end

-- Test 2: io.write log injection
function write_log(message)
    local log_file = io.open("/var/log/app.log", "a")
    -- VULNERABLE: Unsanitized message
    log_file:write(os.date() .. " - " .. message .. "\n")
    log_file:close()
end

-- Test 3: ngx.log injection
function openresty_log()
    local args = ngx.req.get_uri_args()
    local user_id = args.user_id
    -- VULNERABLE: Request param in log
    ngx.log(ngx.ERR, "Processing request for user: " .. user_id)
end

-- Test 4: Multi-line log injection
function log_request_details(request)
    -- VULNERABLE: Newlines can forge log entries
    print("[REQUEST] Path: " .. request.path .. " User: " .. request.user)
end

-- Test 5: Log sensitive data
function log_login_attempt(username, password)
    -- VULNERABLE: Logging password
    print("[AUTH] Login attempt - User: " .. username .. " Password: " .. password)
end

-- Test 6: Error logging with user input
function log_error(error_msg, user_input)
    -- VULNERABLE: User input in error log
    io.stderr:write("ERROR: " .. error_msg .. " - Input: " .. user_input .. "\n")
end

-- Test 7: Debug logging sensitive data
function debug_log(data)
    -- VULNERABLE: Logging entire data structure
    print("[DEBUG] Data: " .. serialize(data))
end

-- Test 8: Access log manipulation
function log_access()
    local args = ngx.req.get_uri_args()
    local client_ip = args.forwarded_ip or ngx.var.remote_addr
    -- VULNERABLE: Spoofable IP in log
    ngx.log(ngx.INFO, "Access from: " .. client_ip)
end

-- Test 9: Format string in log
function formatted_log(template, ...)
    local args = {...}
    -- VULNERABLE: Template could contain malicious format
    print(string.format(template, unpack(args)))
end

-- Test 10: Log forging via headers
function log_request_headers()
    local headers = ngx.req.get_headers()
    local user_agent = headers["User-Agent"]
    -- VULNERABLE: Headers can contain newlines
    ngx.log(ngx.INFO, "User-Agent: " .. (user_agent or "unknown"))
end

-- Test 11: Audit log injection
function audit_log(user, action, resource)
    local log_entry = string.format(
        "[AUDIT] User=%s Action=%s Resource=%s",
        user, action, resource
    )
    -- VULNERABLE: All fields from user input
    print(log_entry)
end

-- Test 12: Exception logging
function log_exception(err, context)
    -- VULNERABLE: Context may contain sensitive data
    print("[EXCEPTION] " .. tostring(err) .. " Context: " .. serialize(context))
end

-- Test 13: SQL query logging
function log_query(query, params)
    -- VULNERABLE: Logging SQL with params
    print("[SQL] " .. query .. " Params: " .. table.concat(params, ", "))
end

-- Helper function (stub)
function serialize(obj)
    if type(obj) == "table" then
        local s = "{ "
        for k, v in pairs(obj) do
            s = s .. tostring(k) .. "=" .. tostring(v) .. " "
        end
        return s .. "}"
    end
    return tostring(obj)
end

File: smoke_tests/cpp/log_injection.cpp
// Log Injection vulnerabilities in C++
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <ctime>
#include <syslog.h>

// Test 1: Direct user input in log
void log_user_action(const std::string& username) {
    // VULNERABLE: username can contain newlines to forge log entries
    std::cout << "[INFO] User " << username << " logged in" << std::endl;
}

// Test 2: Log to file with user input
void write_log(const std::string& message) {
    std::ofstream log("/var/log/app.log", std::ios::app);
    // VULNERABLE: message not sanitized
    log << "[" << time(nullptr) << "] " << message << std::endl;
}

// Test 3: syslog injection
void syslog_message(const std::string& user_input) {
    // VULNERABLE: user_input can contain format specifiers
    openlog("myapp", LOG_PID, LOG_USER);
    syslog(LOG_INFO, "%s", user_input.c_str());
    closelog();
}

// Test 4: Format string in cerr
void log_error(const std::string& error_details) {
    // VULNERABLE: error_details can contain newlines
    std::cerr << "Error: " << error_details << std::endl;
}

// Test 5: HTTP response splitting through logs
void log_request(const std::string& path, const std::string& user_agent) {
    // VULNERABLE: user_agent can contain CRLF
    std::cout << "Request: " << path << " - Agent: " << user_agent << std::endl;
}

// Test 6: Log injection in structured format
void log_json(const std::string& event, const std::string& data) {
    // VULNERABLE: data can break JSON structure
    std::cout << "{\"event\": \"" << event << "\", \"data\": \"" << data << "\"}" << std::endl;
}

// Test 7: Carriage return injection
void log_transaction(const std::string& account_id, double amount) {
    // VULNERABLE: account_id may contain \r\n
    std::stringstream ss;
    ss << "Transaction: Account=" << account_id << " Amount=" << amount;
    std::cout << ss.str() << std::endl;
}

// Test 8: Exception message logging
void log_exception(const std::exception& e) {
    // VULNERABLE: Exception message from untrusted source
    std::cerr << "Exception caught: " << e.what() << std::endl;
}

// Test 9: Multi-line log entry manipulation
void log_audit(const std::string& user, const std::string& action, const std::string& result) {
    std::ofstream audit("/var/log/audit.log", std::ios::app);
    // VULNERABLE: Any of these can contain newlines
    audit << "User: " << user << std::endl;
    audit << "Action: " << action << std::endl;
    audit << "Result: " << result << std::endl;
    audit << "---" << std::endl;
}
